contracts/contracts/token/OUSD.sol

```diff
@@ -445,159 +391,239 @@ contract OUSD is Initializable, InitializableERC20Detailed, Governable {
+                // accounting if they have a 0 balance.
+                balanceOf(_account) == 0
+            ,
+            "Account must be non-rebasing"
+        );
+        RebaseOptions state = rebaseState[_account];
+        // prettier-ignore
+        require(
+            state == RebaseOptions.StdNonRebasing ||
+                state == RebaseOptions.NotSet,
+            "Only standard non-rebasing accounts can opt in"
+        );
 
-        // Increase rebasing credits, totalSupply remains unchanged so no
-        // adjustment necessary
-        _rebasingCredits = _rebasingCredits.add(_creditBalances[_account]);
+        uint256 balance = balanceOf(_account);
 
-        rebaseState[_account] = RebaseOptions.OptIn;
+        // Account
+        rebaseState[_account] = RebaseOptions.StdRebasing;
+        alternativeCreditsPerToken[_account] = 0;
+        _creditBalances[_account] = _balanceToRebasingCredits(balance);
+        // Globals
+        _adjustGlobals(_creditBalances[_account].toInt256(), -balance.toInt256());
 
-        // Delete any fixed credits per token
-        delete nonRebasingCreditsPerToken[_account];
         emit AccountRebasingEnabled(_account);
     }
 
     /**
-     * @dev Explicitly mark that an address is non-rebasing.
+     * @notice The calling account will no longer receive yield
      */
-    function rebaseOptOut() public nonReentrant {
-        require(!_isNonRebasingAccount(msg.sender), "Account has not opted in");
+    function rebaseOptOut() external {
+        _rebaseOptOut(msg.sender);
+    }
 
-        // Increase non rebasing supply
-        nonRebasingSupply = nonRebasingSupply.add(balanceOf(msg.sender));
-        // Set fixed credits per token
-        nonRebasingCreditsPerToken[msg.sender] = _rebasingCreditsPerToken;
+    function _rebaseOptOut(address _account) internal {
+        require(
+            alternativeCreditsPerToken[_account] == 0,
+            "Account must be rebasing"
+        );
+        RebaseOptions state = rebaseState[_account];
+        require(
+            state == RebaseOptions.StdRebasing || state == RebaseOptions.NotSet,
+            "Only standard rebasing accounts can opt out"
+        );
 
-        // Decrease rebasing credits, total supply remains unchanged so no
-        // adjustment necessary
-        _rebasingCredits = _rebasingCredits.sub(_creditBalances[msg.sender]);
+        uint256 oldCredits = _creditBalances[_account];
+        uint256 balance = balanceOf(_account);
 
-        // Mark explicitly opted out of rebasing
-        rebaseState[msg.sender] = RebaseOptions.OptOut;
-        emit AccountRebasingDisabled(msg.sender);
+        // Account
+        rebaseState[_account] = RebaseOptions.StdNonRebasing;
+        alternativeCreditsPerToken[_account] = 1e18;
+        _creditBalances[_account] = balance;
+        // Globals
+        _adjustGlobals(-oldCredits.toInt256(), balance.toInt256());
```

Comment:
🟠 Once resolution gets too low, and rounding kicks in, this global nonrebasing balance set will actually be too low. (this also applies to every other place in the code doing like things)

This is because the credits set to the account's credits will result in a balance greater than the balance set to the globals.

For example, let's say we have one credit for every 2 balance. We try to set a balance of 101. This results in 51 credits for the account. The account now has a balance of 102. We are okay with that. However, we then set the non-rebasing global to 101. It's now too small, and future rebases will be too large. It also breaks withdrawing all non-rebasing funds.

Now the amounts involved here are tiny! More like it would cause formal verification a problem, than anything in actual practice. However, correct accounting is something that is at least worth taking a swing at.

I'm wondering if the correct way to handle this is to have `_balanceToRebasingCredits` return both the credits and the actual balance that would result from using those credits. We then use that actual new balance in the resulting globals change.

Reply:
This is a great point. I am thinking how to implement this elegantly and I have a proposal here: https://github.com/OriginProtocol/origin-dollar/pull/2313

The code cleans up quite nicely if we pass the credits that affect the rebasingSupply to the `_adjustGlobals` function. But the big downside is in the `_adjustAccount` where the exact balance that impacts the rebasingSupply needs to be converted into credits. Need to think more about this one if there is a way to solve this issue with the proposed code

Reply:
Found a way around it which I think is rather elegant and doesn't suffer from any rounding errors.